merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java

```diff
@@ -503,6 +503,14 @@ public SimulationResults computeResults(
                                  serializedTimeline);
   }
 
+  public Duration getTaskDuration(TaskId id){
```

Comment:
I believe an exception would tell us that the scheduler is not working properly which is good. Do we have a general policy for exceptions in Aerie ?

Reply:
Not really; it's historically been a case-by-case justification, with some rather loose philosophy behind each choice.

Broadly, I think of exceptions in three cases: caller-callee contract failures, component integration contract failures, and faulty developer assumptions.

In the first, the caller directly made an illegal request; either the arguments were bad, or the method was called in a program state in which it must not be. (That's the case here.)

In the second, neither the caller nor callee are directly at fault, but the callee is unable to fulfill its side of the contract -- for instance, a network issue occurred behind an interface that doesn't expose the network details. In that case, whichever component brought the caller's component and callee's component into existence, and wired them together, is at fault -- or at least is in the best position to correct for the fault and move on.

In the third, the program itself isn't really at fault. Rather, the developer(s) who built the program expected that a certain scenario would never happen, and it did. These are good candidates for `Error`-subtype exceptions, as we don't normally want these to be caught -- if we did, the fault would have been accounted for at some level. Worst-case, this is treated the same as the previous class, but likely higher-up in the tree of component decomposition, and with a louder error message to the developers.

In all cases, unchecked exceptions are used. Checked exceptions are *much* more like an "alternate return value", and so don't really capture exceptional situations at all. These days, we have `sealed` interfaces that can be used instead to model alternatives, though niche use cases do occasionally still arise.

Reply:
In this case, we should document the contract that `getTaskDuration` (or alternate name recommended) must only be called when the task is known to have terminated; an exception of a specified type will be thrown otherwise.

Reply:
The engine is now throwing an IllegalArgumentException when the task has not finished (or been scheduled).

Reply:
`IllegalStateException` might be more accurate, since it's not the task ID itself that's invalid, but rather the state of the simulation at which you're querying that task ID.

Reply:
I have replaced IllegalArgumentException exceptions with IllegalStateException. By the way, do you prefer me to resolve these conversations when appropriate or you want to do it ?